dot/sync/chain_sync_test.go

```diff
@@ -0,0 +1,1249 @@
+				err: &workerError{
+					err: context.DeadlineExceeded,
+				},
+			},
+		},
+		"res.err.err.Error() == errNoPeers": {
+			chainSyncBuilder: func(ctrl *gomock.Controller, result *worker) chainSync {
+				return chainSync{
+					workerState: newWorkerState(),
+				}
+			},
+			res: &worker{
+				ctx: context.Background(),
+				err: &workerError{
+					err: errNoPeers,
+				},
+			},
+		},
+		"res.err.err.Error() == protocol not supported": {
+			chainSyncBuilder: func(ctrl *gomock.Controller, result *worker) chainSync {
+				mockNetwork := NewMockNetwork(ctrl)
+				mockNetwork.EXPECT().ReportPeer(peerset.ReputationChange{Value: -2147483648,
+					Reason: "Unsupported protocol"},
+					peer.ID(""))
+				return chainSync{
+					workerState: newWorkerState(),
+					network:     mockNetwork,
+				}
+			},
+			res: &worker{
+				ctx: context.Background(),
+				err: &workerError{
+					err: errors.New("protocol not supported"),
+				},
+			},
+		},
+		"no error, no retries": {
+			chainSyncBuilder: func(ctrl *gomock.Controller, result *worker) chainSync {
+				mockWorkHandler := NewMockworkHandler(ctrl)
+				mockWorkHandler.EXPECT().handleWorkerResult(result).Return(result, nil)
+				return chainSync{
+					workerState: newWorkerState(),
+					handler:     mockWorkHandler,
+				}
+			},
+			res: &worker{
+				ctx: context.Background(),
+				err: &workerError{
+					err: errors.New(""),
+				},
+			},
+		},
+		"no error, maxWorkerRetries 2": {
+			chainSyncBuilder: func(ctrl *gomock.Controller, result *worker) chainSync {
+				mockWorkHandler := NewMockworkHandler(ctrl)
+				mockWorkHandler.EXPECT().handleWorkerResult(result).Return(result, nil)
+				return chainSync{
+					workerState: newWorkerState(),
+					handler:     mockWorkHandler,
+				}
+			},
+			maxWorkerRetries: 2,
+			res: &worker{
+				ctx: context.Background(),
+				err: &workerError{
+					err: errors.New(""),
+				},
+			},
+		},
+	}
+	for testName, tt := range tests {
+		tt := tt
+		t.Run(testName, func(t *testing.T) {
+			t.Parallel()
+			ctrl := gomock.NewController(t)
+			sync := tt.chainSyncBuilder(ctrl, tt.res)
+			err := sync.handleResult(tt.res)
+			if tt.err != nil {
+				assert.EqualError(t, err, tt.err.Error())
+			} else {
+				assert.NoError(t, err)
+			}
+		})
+	}
+}
+
+func newTestChainSyncWithReadyBlocks(t *testing.T, readyBlocks *blockQueue) *chainSync {
+	ctrl := gomock.NewController(t)
+	mockBlockState := NewMockBlockState(ctrl)
+	mockBlockState.EXPECT().GetFinalisedNotifierChannel().Return(make(chan *types.FinalisationInfo))
+
+	cfg := &chainSyncConfig{
+		bs:            mockBlockState,
+		readyBlocks:   readyBlocks,
+		pendingBlocks: newDisjointBlockSet(pendingBlocksLimit),
+		minPeers:      1,
+		maxPeers:      5,
+		slotDuration:  defaultSlotDuration,
+	}
+
+	return newChainSync(cfg)
+}
+
+func newTestChainSync(t *testing.T) *chainSync {
+	readyBlocks := newBlockQueue(maxResponseSize)
+	return newTestChainSyncWithReadyBlocks(t, readyBlocks)
+}
```

Comment:
nit pass these a `ctrl *gomock.Controller` instead of `t *testing.T` so it's more explicit this is configuring mocks and stuff.

Otherwise one dev might assume we can set it up in the parent test instead of the subtests